"Leveraging Containers to Increase Developer Productivity" blogpost (#5252)
microsoft/vscode-docs: blogs/2022/03/28/increase-productivity-with-containers.md

---
Order: 72

> TOCTitle: Increase Productivity with Containers

Author: Updated! Made TOC "Moving from local to remote development" for brevity, but kept the "Leveraging containers" part in the page title

PageTitle: Leveraging Containers to Increase Developer Productivity
MetaDescription: Increase productivity by moving your development environment from local, to containers, to the cloud.
Date: 2022-03-28
Author: Olivia Guzzardo
---

# Leveraging Containers to Increase Developer Productivity

March 28, 2022 by Olivia Guzzardo, [@OliviaGuzzardo](https://twitter.com/OliviaGuzzardo)

One of my favorite things that industry professionals like to ask aspiring developers is, “how many lines of code does the average developer write per day?” Most guess in the hundreds or thousands – they’re always shocked to hear that the actual average figure lies in the tens.

Then what are developers doing with the rest of their time? Sure, some of the remaining time is spent on code design and googling “how to center a div in css”, but an inordinate amount is pure overhead – setting up projects, onboarding other developers, and troubleshooting issues that you can’t seem to repro on your own machine.

> The VS Code team has used this as the heart of our research over the years: if we can reduce time spent on overhead, then we can increase productivity time. We have a vision where developers don’t have to keep fighting the same battles over and over again. This means a consistent dev setup that endures never-ending churn of version upgrades, configuration changes, and hardware refreshes.

Author: Great idea!! Added a link

![A shrug emoji saying "It works on my machine"](it-works-on-my-machine.png)

Let’s paint a picture of what local development looks like in the real world. Bobby is so excited because he just landed his first software engineering job! On day one, he shows up ready to code. Instead, his boss hands him a stack of onboarding notes detailing all the steps that he needs to install to get the code up and running. He ends up spending his first full week doing nothing but running install commands, then being bounced from developer to developer to help troubleshoot his build issues, until he can finally run the project on his laptop.

After a little while on the project, Bobby is delivering features like a champ. Then a security vulnerability test alerts his team that they need to update their .NET Core version. Bobby goes through the effort of upgrading the version on his Windows machine, updating the dependencies in their project, and testing any necessary code changes. He pushes his changes up and everything looks good! Or does it?

> As it turns out, a week later his teammate Susie gets back from being out of office and didn’t get the memo about upgrading. When she pulls the latest code, the build fails. After some back and forth figuring out who the heck broke the build while she was gone, she realizes that she still has the outdated version on her machine.

Reviewer: Around this paragraph, it begins to feel like quite a bit of text. Is there any way to add images, diagrams, and/or shorten and combine some paragraphs?

Author: Updated this section to be shorter + rearranged memes

Clearly, a problem emerges here when everything exists solely on a developer’s machine. Your machine may differ wildly from your teammate’s, whether that be from installed versions of dependencies, or from running an entirely different OS. This can lead to a never-ending cycle of configuration nightmares. Even if you get everything in sync with your coworkers, you lose any confidence that you won’t run into more issues when you go to deploy your code.

Now wouldn’t it be nice to regain that confidence by knowing that your development environment was the same as everyone else’s environment and the deployment environment? This brings us to our first advancement in streamlining development productivity: container-based development.

## Containers are everywhere

“Containers” has been a big buzzword in the industry for a while now, so let’s dive into what exactly containers are. To understand the essence of containers, it helps to look at the comparison with a shipping container in the physical world.

> ![A picture of physical shipping containers and a picture of a virtual container diagram, with an equal sign between to show the comparison](container-comparison.png)

Author: yes, there were a few changes that i had forgotten to commit! should be updated now

## Containers for developers

The idea behind container-based development is to introduce containers at the very start of the development process.  Developers can then do all coding and testing in environments that are consistent with both the production environment and other dev environments. Because of this consistency across environments, time spent troubleshooting different code behavior from one environment to another can be virtually eliminated.

This brings us to the concept of a development container: a container that runs a full-featured development environment. A dev container houses its own app and dependencies, such as the required tools, libraries, and runtimes. In the diagram below, you can see that these dependencies exist in the containers and not on the host machine, which means that you can seamlessly switch between tech stacks in no time at all.

![A diagram showing containers with different tech stacks](container-diagram.png)

> Image Source: Development Containers in Education with Visual Studio Code

Author: removed the outdated source line, resized image

> To provide a way to create and connect to dev containers, VS Code unveiled the [Remote -  Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) in 2019. This extension provides the illusion of local development by leveraging the full power of dev containers, all while never leaving the comfort of VS Code. With over 11 million installs of the extension, it got us thinking: what if you could have a dev container that is hosted in the cloud?

Reviewer: "provides the illusion of local development" might sound like users are in the cloud or otherwise remote machine (common misconception). Maybe something like this would help:

```suggestion
To provide a way to create and connect to dev containers, VS Code unveiled the [Remote -  Containers extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.remote-containers) in 2019. This extension enhances local development by leveraging the full power of dev containers, all while never leaving the comfort of VS Code. With over 11 million installs of the extension, it got us thinking: what if you could have a dev container that is hosted in the cloud?
```

## Moving to the cloud

Let’s be real: everything is moving to the cloud, so why should your development environment be any different? Well, for one, the cloud is vulnerable to attacks. For two, it seems like we’re constantly hearing about major outages. For three, it can be pricey. For four… wait, isn’t the cloud supposed to be a good thing?

With all these reasons to be weary of the cloud, it can be uncomfortable to think about something as important as our development environment being hosted there.  So let’s stay comfortable! Let’s continue relying on our 6-year-old laptop that makes really weird whirring noises if you try to open an email too quickly, let alone try to build a project. And we’ll just comfortably wait for the inevitability of when the laptop crashes, and then once we get our new laptop, we have to rebuild our dev environment and try to remember exactly how we had it configured.

![A meme showing a dog saying "This is fine" while sitting in a room on fire](this-is-fine-meme.png)

Turns out staying in our comfort zone doesn’t sound so peachy either.

> While there are inherent risks with the cloud, it also comes with major benefits: scalability, faster performance, and easier maintenance, just to name a few. Hosting in the cloud gives you access to another machine’s resources that can then be spun up quickly and available anywhere. Combine that with the benefits of container-based development, and developers can be up and coding in the blink of an eye.

Reviewer: Rather than implying we just have to accept those risks from cloud-based services in the first paragraph, could we also mention how Codespaces tries to avoid them, further increasing users' confidence?

Author: great point! i added a couple paragraphs about this in the next section when i talk about codespaces - let me know if you think it would benefit to add in this section too though!

Reviewer: Would you be able to point to which paragraphs address this? Want to make sure I'm not missing it!

Author: tagged you in the section :)

## Containers in the cloud

> Running containers in the cloud is not a new concept; in fact, [one study](https://www.datadoghq.com/container-report/?utm_source=SocialMedia&utm_medium=Twitter&utm_campaign=OrganicPosting-containerreport) shows that at least half of all containers are running in the cloud. The basic infrastructure involves deploying a container to a cloud-hosted VM. For a development environment hosted in the cloud, the deployed container is the dev container that houses our code and dependencies. Developers can then connect to this VM to harness the power of containers in the cloud.

Reviewer: The last two sentences here are a bit confusing to me. Maybe something like this would help:

```suggestion
Running containers in the cloud is not a new concept; in fact, [one study](https://www.datadoghq.com/container-report/?utm_source=SocialMedia&utm_medium=Twitter&utm_campaign=OrganicPosting-containerreport) shows that at least half of all containers are running in the cloud. The basic infrastructure involves deploying a container to a cloud-hosted VM. We can deploy our dev container to the cloud  to provide a cloud-hosted development environment.
```

VS Code’s entry into this space comes from powering [GitHub Codespaces](https://github.com/features/codespaces). Within minutes, you can create and configure a dev container hosted in the cloud that is ready whenever you need it. You can then connect via VS Code (in either the browser or desktop) to a development environment that is managed completely for you, no longer relying on your laptop’s resources to handle the demand.

![A diagram showing the infrastructure and workflow of Codespaces](codespaces-diagram.png)

> Image Source: GitHub Codespaces overview - GitHub Docs

Reviewer: Should this be a link?

Also, it feels like the Codespaces section is surprisingly shorter than the rest. Perhaps a screenshot of a Codespace would be helpful to add more images and balance to the post?

Also, there isn't an explicit mention that the same dev containers can be used in Remote - Containers and Codespaces, providing a truly seamless transition experience. Some users may want Remote - Containers (prefer desktop, no cost, can work offline) while other like Codespaces (harness remote compute, don't have Docker installed).

Author: Added the explicit mention about reusing dev containers, and added more detail about how github codespaces alleviates previously-mentioned concerns about the cloud to balance out this section

## What’s next?

> As developers, we want to spend more time developing software and less time on headache-inducing configuration. Trends in the industry can, and should, be used to empower developers to increase their productivity. We’ve seen how containers and the cloud can take us to the next level – what do you think could be the next step in making our lives even easier?

Reviewer: Do we want a CTA, i.e. check out Remote - Containers extension or doc for more?

Author: added CTA to the getting started documentation for remote containers
